text/2021-12-29-openapi-to-typescript-api-client-and-forms.md

```diff
@@ -0,0 +1,458 @@
+# OpenAPI to TypeScript API Client and Forms
+
+## Summary
+
+Use the OpenAPI specification to generate a TypeScript API client and forms.
+
+## Motivation
+
+Ref: https://github.com/chaos-mesh/chaos-mesh/issues/2615.
+
+Currently, the chaos types are split between the front-end and back-end, and we
+can‘t reuse types that have already been defined.
+
+You can find these **extra** type definitions in https://github.com/chaos-mesh/chaos-mesh/blob/release-2.1/ui/app/src/components/NewExperimentNext/data/types.ts.
+
+This leads to several problems:
+
+- The front-end needs to be manually synchronized when there is an update to the
+  type definition.
+- As `1` continues to emerge, there are more and more things to maintain manually.
+
+Considering our existing maintenance cost and past experience, we are not able
+to synchronize the corresponding descriptions to the front-end in a timely manner.
+
+This also works against those who want to contribute to Chaos Mesh, and it would
+be great if only modifying the API will synchronize the changes to the UI.
+
+So the best solution for now is to automate the generation of the schemas needed
+for the front-end.
+
+## Detailed design
+
+I will split the details into two parts:
+
+- The first part is the generation of the TypeScript schemas.
+- The second part is how to use the generated files to produce a forms skeleton.
+
+### Generate TypeScript Schemas
+
+Luckily, there has serval tools can help us to generate the schemas, all of them
+has pros and cons, finally I choose the
+[OpenAPITools/openapi-generator](https://github.com/OpenAPITools/openapi-generator),
+because:
+
+- Pros
+  - It's official.
+  - It's a very popular tool and has a huge community.
+- Cons
+  - The nodejs package is just a wrapper of jar, so JRE is needed.
```

Comment:
Is JRE only needed when building, and don't need to install it in the image?

Reply:
Yep. Only needed in build phase.